### DISCUSSION TITLE
Form validation rules outside of schema

### DISCUSSION THREAD

**Discussion by helloreece** [filamentphp/filament#13325] [Help]
### Package

Panel builder

### Package Version

3.2.x

### How can we help you?

Hi,

In my application, models are managed via a mix of Filament Forms, console tasks, and an API. It's becoming a fairly large application to manage.

Working with my models outside of just the standard http request has made me ponder whether the Filament Form schema() is the best place for model validation rules?

Is there an approach for using Filament Forms whereby validation rules can be set more globally, and the form method utilises that.

Full disclosure... I'm a year into my programming journey. This may not even make sense as a question 😂

**Comment by hotsaucejake** [comment]
Are you talking about how many of the items have the same validation when you talk about "globally?"  You could make a trait for this.

**Comment by helloreece** [comment]
Maybe an example will illustrate things better.

I have a Contact model.

My Filament ContactResource defines the form schema for creating a new
contact. It includes validation, such as required() and other rules.

However, there are other places in my application that need to create a
Contact. For example, actions from console commands, API requests, etc.

I can use Spatie Laravel Data package to make a centralised ContactData
definition. This includes a method called rules() for the validation rules
which should be applied.

This solves the problem mostly, but my Filament Form still requires its own
validation rules to be added to the form schema.

It means all my Filament Resources have repeated validation logic.

Any best practice or advice would be appreciated.

Many thanks 🙏

**Comment by hotsaucejake** [comment]
For being a year into programming, it's good that you don't want to repeat code and looking for ways to centralize logic.  Although this doesn't pertain to filament per se, you could take an approach of a static function within your model itself.

Within your contact model you could great a public static function getValidations() which returns and array of your validations.  So in your form schema you could do something like ->rules(Contact::getValidations()).  If you have form requests you could re-use this logic.

As your app progresses you can alter the function according to different uses.  For example, if your validation is slightly different for something else you could still use the same function but have different parameters set.  

public static function getValidations($forSomethingElse = false) {
if ($forSomethingElse) {
// return a different array
} else {
// return original array
}
}

**Comment by helloreece** [comment]
This is a helpful thought and something I will consider for sure. Thank you 🙏

I think for more complex use cases taking things outside of the panel builder makes things easier to achieve. This is where I'm leaning currently. 

If anyone reads this and has further thoughts, do share them in case others find this thread.

**Comment by fritzfr** [comment]
This is an interesting question because I also see others stumbling across this issue and I did so myself (https://github.com/filamentphp/filament/discussions/16038). This post here is also related https://github.com/filamentphp/filament/discussions/4227.

Validation rules do not belong into the schema, it's not easily readable and a bad pattern.

**Comment by pabbec** [comment]
Hi! I'm having the same issue. It's not a problem, but it's something that has to do with good practices. Filament is extraordinary, and the way it validates is correct from the perspective it's presented.

I'm testing, and even using an example of the Contact model that could be created from Filament as a resource, from a command, or an API. What I did was:

- I have an action class that manages the creation of the Contact:
``\App\Actions\Crm\Contact\CreateContactAction.php``
I can invoke this action from my Filament Resource, a command, or an API.
- I have a class (it could well be in the model) called ContactRules where I define an array $validations_rules with all the field-by-field validation rules. I also have two methods: getValidationRules() and getValidationRulesForField($field_name) . I think they're pretty self-explanatory. E.g.:

```// ContactRules.php
class ContactRules
{
    protected static string $upload_disk_id = 'cdn';
    protected static string $upload_directory_name = 'crm/contacts';

    protected static $validation_rules = [
        'name' => ['required','string','min:2','max:50'],
        'email' => ['nullable','email'],
        'phone' => ['nullable','min:13', 'max:16'],
        'is_visible' => ['required','boolean'],
    ];
    ...
```
- In this class, I also add two other methods: storageRules , where I save the disk and directory where the files should be saved (let's assume, in this case, the avatar).
- I also use the Spatie Laravel Data package to use DTOs.
```
// CreateContactDto.php
use Spatie\LaravelData\Data;
use App\Models\Crm\Contact;
use App\Models\Rules\Crm\ContactRules.php
use Spatie\LaravelData\Support\Validation\ValidationContext;

class CreateContactDto extends Data
{ 
        public function __construct( 
                public null|int $id = null, 
                public string $name, 
                public ?string $email, 
                public ?string $phone, 
                public bool $is_visible = true, 
        ) {} 


        public static function rules(): array 
        { 
              return ContactRules::getValidationRules(); 
        } 
        
        public static function attributes(): array 
        { 
              return ContactRules::getAttributeNames(); 
        }
}
```
Then, in ``Resources\ContactsResources\Pages\CreateContact.php``, I use the ``handleRecordCreation(array $data)`` method, which receives the array with all the form data.

Here I do the following:
``
// CreateContact.php
try {
        $dto = CreateContactDTO::validateAndCreate($data);
        $contact = CreateContactAction::run($dto);
        return $category;
}
catch (ValidationException $e) {
        Notification::make()
        ->title('Validation Error')
        ->body('Review the form and correct any errors.')
        ->danger()
        ->send();
        throw ValidationException::withMessages(
            collect($e->errors())
                ->mapWithKeys(fn($msgs, $key) => ["data.{$key}" => $msgs[0]])
                ->toArray()
        );
}
``
When creating the DTO, the model is automatically validated. Otherwise, this triggers a ValidateException exception, which I catch, send a notification, and then trigger again, transforming the error array mapped with ``field => message`` to ``data.field => message`` so that they are displayed on the form.

So far, this works perfectly. I can even include validation rules in the form if I want:
```
TextInput::make('name')
->label('Name')
->columnSpanFull()
->rules(ContactRules::getFieldValidationRules('name'))
```
But my problem is now when I want to do this with a modal... I can't get the error messages to display for each field :(

**Reply by leandrocfe** [reply]
I think Actions use  `mountedActionsData.0.` instead of `data.` as StatePath

**Comment by pabbec** [comment]
You right! thanks!
